tripal/tripal
# Allow fields to link to their corresponding entity

# New Feature

### Closes #1632 

### ~~Dependent on #1841~~ (merged)
### ~~Dependent on #1854~~ (merged)

### Tripal Version: 4.x

## Description
None of the Tripal 4 fields currently are "clickable" like they were in Tripal 3.
This PR will eventually make this possible.  For example
![2024-02-26_organism-with-link](https://github.com/tripal/tripal/assets/8419404/6e8f2f8d-082f-46bc-9507-460cdaa98a2b)

## Tasks
 - 🙈 ~~Create a new service to generate the entity titles for either one or all records. For example~~
```
    $title_manager = \Drupal::service('tripal.tripal_entity.title');
    $titles = $title_manager->getEntityTitles($datastore, $bundle, $record_id);
```
 ~~This will be a chunk of the code from the publish service, so make it generic enough that the publish service can also use it.~~

 - 🙈 ~~Update the publish service to use the new `tripal.entity_title` service.
 To implement this, make it a child class of the class above.~~
```
use \Drupal\tripal\Services\TripalEntityTitle;
class TripalPublish extends TripalEntityTitle {`
```

 - Abandoned these first two tasks, initial implementation is very slow at ~1.8 seconds to return the url. Generating a title has a large amount of overhead for setup. This was okay for publish, but not for this. Implement an alternative method.

 - [x] Create a new service `tripal.tripal_entity.lookup` to generate either the entity ID or the local URL for a record.
~~This will internally call the `tripal.tripal_entity.title` service.~~

 - [x] Also in this service implement a way to look up a bundle id from a CV term. Such as, if the field links to a gene, we need to know that the bundle name is gene. We can get this from the CV term, which in the field formatter can be retrieved with
```
  $item_settings = $item->getDataDefinition()->getSettings();
  $termIdSpace = $item_settings['termIdSpace'];
  $termAccession = $item_settings['termAccession'];
```
 - [x] Create a function to look up not-null columns in a chado table `getNotNullColumns()`
This will be the only slow step, but we can cache the result with `\Drupal::cache()->set($cache_id, $cache_values);` We only need a single cached set of values for each base table that shows up as a content type, and we don't even need to worry about expiring it after some time because this information won't change.

The reason we want this information is that the chado table and column for the corresponding field is guaranteed to be populated for every entity, because of the non-null constraint.

 - [x] Implement the entity lookup as follows
Select one of the entity tables with a not-null constraint. This insures there is a row for every published entity. Since we know the chado record_id, we can then look for the corresponding entity id with a query similar to this
```
$required_fields = $this->getRequiredFields($bundle_id, $entity_type);
$required_field = array_key_first($required_fields);
$pkey_property_id = 'record_id';  // This assumes all of our fields use this id - so far this is true
$required_property = $required_field . '.' . $pkey_property_id;
$query = \Drupal::entityQuery($entity_type)
      ->condition('type', $bundle_id)
      ->condition($required_property, $record_id, '=')
      ->accessCheck(TRUE);
// The values of the array are always entity ids. The keys will be
// revision ids if the entity supports revision and entity ids if not.
$ids = $query->execute();
```

 - [x] A high level function to simplify code in the  field formatters
~~public function getFieldUrl($displayed_string, $record_id, $item_settings)~~
`$renderable_item = $lookup_manager->getRenderableItem($displayed_string, $values['entity_id']);`

 - [x] Entity lookup is now implemented in the tripalTypes function for each field using a function callback.
```
    // This property will store the Drupal entity ID of the linked chado
    // record, if one exists.
    $properties[] = new ChadoIntStoragePropertyType($entity_type_id, self::$id, 'entity_id', self::$drupal_entity_term, [
      'action' => 'function',
      'drupal_store' => TRUE,
      'namespace' => self::$chadostorage_namespace,
      'function' => self::$drupal_entity_callback,
      'fkey' => self::$object_id,
    ]);
```
The field should then have clickable links, for example
![2024-02-13_issue1632](https://github.com/tripal/tripal/assets/8419404/3076a637-d1cd-45d3-a900-113724808c9d)

 - [x] Handle the cases where the chado base table does not exactly match the drupal table name derived from the content type name, e.g. the `arraydesign` table contact reference is in the table `tripal_entity__array_design_manufacturer` because the content type id is `array_design`
 This also happens for any of the feature table content types, analysis (genome_assembly), etc.

- [x] Figure out URL generation - ~~currently hardcoded because of multi-second time required.~~
The solution was this `$url_object = Url::fromRoute('entity.tripal_entity.canonical', ['tripal_entity' => $entity_id]);` which takes only a few milliseconds.

 - [x] Move the entity lookup to the tripalTypes field functions. The entity will now be looked up when creating content and the formatter will use this value to generate the link.

 - [x] Create tests for the new service

 - [x] Create an update function to update the field tables with the new entity id property

## Testing?
 1. Import all the content type groups
 2. Start adding content, and as you go link to as much as possible. This order works well for content types that have dependencies: Contact, Project, Publication, Protocol, Array Design, Assay, Analysis, Biological Sample, Germplasm Accession, Genetic Marker, Genetic Map, and whatever else seems good.
Note on the Array Design page that the "Manufacturer" is a foreign key to contact, so we should get a link displayed.
Note on the Assay page that the "Operator" is also a foreign key to contact, however the term for this one is NCIT:C48036 which is not a content type, so no link will be added. 🙈 Likewise for Biosource Provider on Biomaterial content type...
 3. Make sure this field works
```
    -   name: analysis_software
        content_type: analysis
        label: Program, Pipeline, Workflow or Method Name
            storage_plugin_settings:
                base_table: analysis
                base_column: program
```

 4. To test the update function, 
   4a. Build a new docker on the 4.x branch
   4b. Add a project and add a contact (without any links)
   4c. Make a link `sitedb=> insert into project_contact (project_id, contact_id) values (1,2);`
   4d. Switch branch `git checkout tv4g1-issue1632-add-field-links-to-entities`
   4e. Run updates `drush updatedb`
   4f. Publish "Project" and "Contact" content types
   4g. run the jobs
   4h. rebuild caches
   4i. Go to your project and contact and confirm that the field is a link to the other e.g.
   ![2024-04-17_project](https://github.com/tripal/tripal/assets/8419404/4ee3641a-36da-471b-934b-b647a0befbd5)
